File: backend/python_processor/exercises/fist_exercise.py
import cv2
import numpy as np
from .base_exercise import BaseExercise

class FistExercise(BaseExercise):
    """Упражнение: Кулак (все пальцы сжаты)"""

    def __init__(self):
        super().__init__()
        self.name = "Кулак"
        self.description = "Сожмите все пальцы в кулак"
        self.exercise_id = "fist"

    def check_fingers(self, finger_states, hand_landmarks, frame_shape):
        """Проверяет, сжаты ли все пальцы"""
        raised_fingers = sum(finger_states)

        # Для кулака должно быть поднято не более 1 пальца
        is_correct = raised_fingers <= 1

        if is_correct:
            message = "✅ Кулак сжат правильно!"
        else:
            message = f"❌ Сожмите пальцы (поднято {raised_fingers})"

        return is_correct, message

    def get_finger_colors(self, finger_states):
        """Цвета для кулака:
        - Зеленый если палец сжат
        - Красный если палец поднят
        """
        colors = []
        for is_raised in finger_states:
            if is_raised:
                colors.append((0, 0, 255))  # Красный - ошибка (палец поднят)
            else:
                colors.append((0, 255, 0))  # Зеленый - правильно (палец сжат)
        return colors

File: LF/app/src/main/java/com/example/lf/exercises/FistExercise.kt
package com.example.lf.exercises

import com.google.mediapipe.tasks.components.containers.NormalizedLandmark

class FistExercise : BaseExercise() {

    companion object {

        private const val DEFAULT_TOTAL_CYCLES = 10
        private const val DEFAULT_HOLD_DURATION_MS = 1500L

        private const val STATE_WAITING_FIST = "waiting_fist"
        private const val STATE_HOLDING_FIST = "holding_fist"

        private const val MAX_RAISED_FINGERS_FOR_FIST = 1
    }

    init {
        name = "Кулак"
        description = "Сжимайте и разжимайте кулаки для укрепления кистей"
        exerciseId = "fist"
        totalCycles = DEFAULT_TOTAL_CYCLES
        holdDuration = DEFAULT_HOLD_DURATION_MS
    }

    override fun reset() {
        super.reset()
        state = STATE_WAITING_FIST
        currentCycle = 0
        completed = false
        calibrated = true

        autoReset = false
    }

    override fun checkFingers(
        fingerStates: List<Boolean>,
        landmarks: List<NormalizedLandmark>,
        frameWidth: Int,
        frameHeight: Int
    ): Pair<Boolean, String> {

        if (autoReset && completed) {
            reset()
            autoReset = false
        }

        if (completed) {
            return Pair(true, "🎉 Упражнение завершено! Отличная работа!")
        }

        val raisedFingersCount = fingerStates.count { it }
        val isFist = raisedFingersCount <= MAX_RAISED_FINGERS_FOR_FIST
        val currentTime = System.currentTimeMillis()

        when (state) {
            STATE_WAITING_FIST -> {
                if (isFist) {
                    state = STATE_HOLDING_FIST
                    holdStart = currentTime
                }
            }
            STATE_HOLDING_FIST -> {
                if (!isFist) {

                    state = STATE_WAITING_FIST
                } else if (currentTime - holdStart >= holdDuration) {

                    currentCycle++
                    cycleCompleted = true
                    state = STATE_WAITING_FIST

                    if (currentCycle >= totalCycles) {
                        completed = true
                        autoReset = true
                    }
                }
            }
        }

        val message = buildMessage(raisedFingersCount, currentTime)
        val isHolding = (state == STATE_HOLDING_FIST)

        return Pair(isHolding, message)
    }

    private fun buildMessage(raisedFingersCount: Int, currentTime: Long): String {
        return when {
            completed -> "🎉 Упражнение завершено! Отличная работа!"
            state == STATE_HOLDING_FIST -> {
                val elapsed = currentTime - holdStart
                val remainingSeconds = ((holdDuration - elapsed) / 1000).toInt() + 1
                "Держите кулак... ${remainingSeconds}с (${currentCycle + 1}/$totalCycles)"
            }
            else -> {
                val hint = if (raisedFingersCount > MAX_RAISED_FINGERS_FOR_FIST) {
                    " (поднято пальцев: $raisedFingersCount, нужно ≤ $MAX_RAISED_FINGERS_FOR_FIST)"
                } else ""
                "Сожмите кулак$hint (${currentCycle + 1}/$totalCycles)"
            }
        }
    }

    override fun getFingerColors(fingerStates: List<Boolean>): List<Int> {
        val isHolding = state == STATE_HOLDING_FIST
        val colors = mutableListOf<Int>()

        for (isRaised in fingerStates) {
            val color = when {

                isHolding && !isRaised -> COLORS["green"]!!

                !isHolding && !isRaised -> COLORS["cyan"]!!

                else -> COLORS["red"]!!
            }
            colors.add(color)
        }
        return colors
    }
}